# Exercise 02 -  OLAP Cubes - Grouping Sets

All the databases table in this demo are based on public database samples and transformations
- `Sakila` is a sample database created by `MySql` [Link](https://dev.mysql.com/doc/sakila/en/sakila-structure.html)
- The postgresql version of it is called `Pagila` [Link](https://github.com/devrimgunduz/pagila)
- The facts and dimension tables design is based on O'Reilly's public dimensional modelling tutorial schema [Link](http://archive.oreilly.com/oreillyschool/courses/dba3/index.html)

Start by connecting to the database by running the cells below. If you are coming back to this exercise, then uncomment and run the first cell to recreate the database. If you recently completed the slicing and dicing exercise, then skip to the second cell.

In [1]:
!PGPASSWORD=student createdb -h 127.0.0.1 -U student pagila_star
!PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila_star -f Data/pagila-star.sql

createdb: database creation failed: ERROR:  database "pagila_star" already exists
 set_config 
------------
 
(1 row)

psql:Data/pagila-star.sql:41: ERROR:  type "mpaa_rating" already exists
psql:Data/pagila-star.sql:51: ERROR:  type "year" already exists
psql:Data/pagila-star.sql:68: ERROR:  function "_group_concat" already exists with same argument types
psql:Data/pagila-star.sql:85: ERROR:  function "film_in_stock" already exists with same argument types
psql:Data/pagila-star.sql:102: ERROR:  function "film_not_in_stock" already exists with same argument types
psql:Data/pagila-star.sql:147: ERROR:  function "get_customer_balance" already exists with same argument types
psql:Data/pagila-star.sql:169: ERROR:  function "inventory_held_by_customer" already exists with same argument types
psql:Data/pagila-star.sql:206: ERROR:  function "inventory_in_stock" already exists with same argument types
psql:Data/pagila-star.sql:224: ERROR:  function "last_day" already exists with same argument 

### Connect to the local database where Pagila is loaded

In [2]:
import sql
%load_ext sql

DB_ENDPOINT = "127.0.0.1"
DB = 'pagila_star'
DB_USER = 'student'
DB_PASSWORD = 'student'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)
%sql $conn_string

postgresql://student:student@127.0.0.1:5432/pagila_star


'Connected: student@pagila_star'

### Star Schema

<img src="pagila-star.png" width="50%"/>

# Grouping Sets
- It happens often that for 3 dimensions, you want to aggregate a fact:
    - by nothing (total)
    - then by the 1st dimension
    - then by the 2nd 
    - then by the 3rd 
    - then by the 1st and 2nd
    - then by the 2nd and 3rd
    - then by the 1st and 3rd
    - then by the 1st and 2nd and 3rd
    
- Since this is very common, and in all cases, we are iterating through all the fact table anyhow, there is a more clever way to do that using the SQL grouping statement "GROUPING SETS" 

## Total Revenue

TODO: Write a query that calculates total revenue (sales_amount)

In [3]:
%%time
%%sql
SELECT SUM(sales_amount) AS revenue
FROM factsales
JOIN dimdate ON factsales.date_key = dimdate.date_key
JOIN dimmovie ON factsales.movie_key = dimmovie.movie_key
JOIN dimcustomer ON factsales.customer_key = dimcustomer.customer_key

 * postgresql://student:***@127.0.0.1:5432/pagila_star
1 rows affected.
CPU times: user 3.98 ms, sys: 0 ns, total: 3.98 ms
Wall time: 20.5 ms


[(Decimal('67416.51'),)]

## Revenue by Country
TODO: Write a query that calculates total revenue (sales_amount) by country

In [4]:
%%time
%%sql
SELECT dimcustomer.country, SUM(sales_amount) AS revenue
FROM factsales
JOIN dimdate ON factsales.date_key = dimdate.date_key
JOIN dimmovie ON factsales.movie_key = dimmovie.movie_key
JOIN dimcustomer ON factsales.customer_key = dimcustomer.customer_key
GROUP BY (dimcustomer.country)
ORDER BY revenue DESC
LIMIT 20

 * postgresql://student:***@127.0.0.1:5432/pagila_star
20 rows affected.
CPU times: user 4.06 ms, sys: 554 µs, total: 4.61 ms
Wall time: 26.6 ms


[('India', Decimal('6630.27')),
 ('China', Decimal('5802.73')),
 ('United States', Decimal('4110.32')),
 ('Japan', Decimal('3471.74')),
 ('Mexico', Decimal('3307.04')),
 ('Brazil', Decimal('3200.52')),
 ('Russian Federation', Decimal('3045.87')),
 ('Philippines', Decimal('2381.32')),
 ('Turkey', Decimal('1662.12')),
 ('Nigeria', Decimal('1511.48')),
 ('Indonesia', Decimal('1510.33')),
 ('Argentina', Decimal('1434.48')),
 ('Taiwan', Decimal('1210.94')),
 ('South Africa', Decimal('1204.15')),
 ('Iran', Decimal('950.75')),
 ('United Kingdom', Decimal('924.80')),
 ('Poland', Decimal('877.97')),
 ('Italy', Decimal('831.11')),
 ('Germany', Decimal('831.04')),
 ('Vietnam', Decimal('746.28'))]

## Revenue by Month
TODO: Write a query that calculates total revenue (sales_amount) by month

In [5]:
%%time
%%sql
SELECT dimdate.month, SUM(sales_amount) AS revenue
FROM factsales
JOIN dimdate ON factsales.date_key = dimdate.date_key
JOIN dimmovie ON factsales.movie_key = dimmovie.movie_key
JOIN dimcustomer ON factsales.customer_key = dimcustomer.customer_key
GROUP BY (dimdate.month)
ORDER BY revenue DESC
LIMIT 20

 * postgresql://student:***@127.0.0.1:5432/pagila_star
5 rows affected.
CPU times: user 3.44 ms, sys: 477 µs, total: 3.91 ms
Wall time: 21.6 ms


[(4, Decimal('28559.46')),
 (3, Decimal('23886.56')),
 (2, Decimal('9631.88')),
 (1, Decimal('4824.43')),
 (5, Decimal('514.18'))]

## Revenue by Month & Country
TODO: Write a query that calculates total revenue (sales_amount) by month and country. Sort the data by month, country, and revenue in descending order. The first few rows of your output should match the table below.

In [6]:
%%time
%%sql
SELECT dimdate.month, dimcustomer.country, SUM(sales_amount) AS revenue
FROM factsales
JOIN dimdate ON factsales.date_key = dimdate.date_key
JOIN dimmovie ON factsales.movie_key = dimmovie.movie_key
JOIN dimcustomer ON factsales.customer_key = dimcustomer.customer_key
GROUP BY (dimdate.month,dimcustomer.country)
ORDER BY dimdate.month, dimcustomer.country, revenue DESC
LIMIT 20

 * postgresql://student:***@127.0.0.1:5432/pagila_star
20 rows affected.
CPU times: user 5.17 ms, sys: 0 ns, total: 5.17 ms
Wall time: 27.5 ms


[(1, 'Algeria', Decimal('33.92')),
 (1, 'Angola', Decimal('27.93')),
 (1, 'Anguilla', Decimal('6.97')),
 (1, 'Argentina', Decimal('81.78')),
 (1, 'Austria', Decimal('22.92')),
 (1, 'Azerbaijan', Decimal('20.97')),
 (1, 'Bahrain', Decimal('3.99')),
 (1, 'Bangladesh', Decimal('19.95')),
 (1, 'Belarus', Decimal('5.98')),
 (1, 'Bolivia', Decimal('4.97')),
 (1, 'Brazil', Decimal('210.50')),
 (1, 'Brunei', Decimal('5.99')),
 (1, 'Bulgaria', Decimal('9.98')),
 (1, 'Cambodia', Decimal('11.96')),
 (1, 'Cameroon', Decimal('13.97')),
 (1, 'Canada', Decimal('33.93')),
 (1, 'Chad', Decimal('12.96')),
 (1, 'Chile', Decimal('24.95')),
 (1, 'China', Decimal('455.91')),
 (1, 'Colombia', Decimal('47.87'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>month</th>
        <th>country</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>1</td>
        <td>Australia</td>
        <td>2364.19</td>
    </tr>
    <tr>
        <td>1</td>
        <td>Canada</td>
        <td>2460.24</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Australia</td>
        <td>4895.10</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Canada</td>
        <td>4736.78</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Australia</td>
        <td>12060.33</td>
    </tr>
</tbody></table></div>

## Revenue Total, by Month, by Country, by Month & Country All in one shot

TODO: Write a query that calculates total revenue at the various grouping levels done above (total, by month, by country, by month & country) all at once using the grouping sets function. Your output should match the table below.

In [7]:
%%time
%%sql
SELECT dimdate.month, dimcustomer.country, SUM(sales_amount) AS revenue
FROM factsales
JOIN dimdate ON factsales.date_key = dimdate.date_key
JOIN dimmovie ON factsales.movie_key = dimmovie.movie_key
JOIN dimcustomer ON factsales.customer_key = dimcustomer.customer_key
GROUP BY GROUPING SETS ((), dimdate.month, dimcustomer.country, 
                         (dimdate.month, dimcustomer.country))
ORDER BY dimdate.month, dimcustomer.country, revenue DESC
LIMIT 20

 * postgresql://student:***@127.0.0.1:5432/pagila_star
20 rows affected.
CPU times: user 6.37 ms, sys: 41 µs, total: 6.42 ms
Wall time: 50.4 ms


[(1, 'Algeria', Decimal('33.92')),
 (1, 'Angola', Decimal('27.93')),
 (1, 'Anguilla', Decimal('6.97')),
 (1, 'Argentina', Decimal('81.78')),
 (1, 'Austria', Decimal('22.92')),
 (1, 'Azerbaijan', Decimal('20.97')),
 (1, 'Bahrain', Decimal('3.99')),
 (1, 'Bangladesh', Decimal('19.95')),
 (1, 'Belarus', Decimal('5.98')),
 (1, 'Bolivia', Decimal('4.97')),
 (1, 'Brazil', Decimal('210.50')),
 (1, 'Brunei', Decimal('5.99')),
 (1, 'Bulgaria', Decimal('9.98')),
 (1, 'Cambodia', Decimal('11.96')),
 (1, 'Cameroon', Decimal('13.97')),
 (1, 'Canada', Decimal('33.93')),
 (1, 'Chad', Decimal('12.96')),
 (1, 'Chile', Decimal('24.95')),
 (1, 'China', Decimal('455.91')),
 (1, 'Colombia', Decimal('47.87'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>month</th>
        <th>country</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>1</td>
        <td>Australia</td>
        <td>2364.19</td>
    </tr>
    <tr>
        <td>1</td>
        <td>Canada</td>
        <td>2460.24</td>
    </tr>
    <tr>
        <td>1</td>
        <td>None</td>
        <td>4824.43</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Australia</td>
        <td>4895.10</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Canada</td>
        <td>4736.78</td>
    </tr>
    <tr>
        <td>2</td>
        <td>None</td>
        <td>9631.88</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Australia</td>
        <td>12060.33</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Canada</td>
        <td>11826.23</td>
    </tr>
    <tr>
        <td>3</td>
        <td>None</td>
        <td>23886.56</td>
    </tr>
    <tr>
        <td>4</td>
        <td>Australia</td>
        <td>14136.07</td>
    </tr>
    <tr>
        <td>4</td>
        <td>Canada</td>
        <td>14423.39</td>
    </tr>
    <tr>
        <td>4</td>
        <td>None</td>
        <td>28559.46</td>
    </tr>
    <tr>
        <td>5</td>
        <td>Australia</td>
        <td>271.08</td>
    </tr>
    <tr>
        <td>5</td>
        <td>Canada</td>
        <td>243.10</td>
    </tr>
    <tr>
        <td>5</td>
        <td>None</td>
        <td>514.18</td>
    </tr>
    <tr>
        <td>None</td>
        <td>None</td>
        <td>67416.51</td>
    </tr>
    <tr>
        <td>None</td>
        <td>Australia</td>
        <td>33726.77</td>
    </tr>
    <tr>
        <td>None</td>
        <td>Canada</td>
        <td>33689.74</td>
    </tr>
</tbody></table></div>